src/c++/perf_analyzer/test_request_rate_manager.cc

```diff
@@ -39,251 +39,295 @@ namespace triton { namespace perfanalyzer {
+        CHECK(stats->num_async_stream_infer_calls > 0);
+        CHECK(stats->num_start_stream_calls > 0);
+      } else {
+        CHECK(stats->num_infer_calls == 0);
+        CHECK(stats->num_async_infer_calls > 0);
+        CHECK(stats->num_async_stream_infer_calls == 0);
+        CHECK(stats->num_start_stream_calls == 0);
       }
-      else {
-        throw std::invalid_argument("Unexpected distribution type");
+    } else {
+      if (params.streaming) {
+        CHECK(stats->num_infer_calls > 0);
+        CHECK(stats->num_async_infer_calls == 0);
+        CHECK(stats->num_async_stream_infer_calls == 0);
+        CHECK(stats->num_start_stream_calls > 0);
+      } else {
+        CHECK(stats->num_infer_calls > 0);
+        CHECK(stats->num_async_infer_calls == 0);
+        CHECK(stats->num_async_stream_infer_calls == 0);
+        CHECK(stats->num_start_stream_calls == 0);
       }
     }
+  }
+
+  void CheckCallDistribution(
+      Distribution request_distribution, int request_rate)
+  {
+    auto timestamps = GetStats()->request_timestamps;
+    std::vector<int64_t> time_delays = GatherTimeBetweenRequests(timestamps);
+
+    double delay_average = CalculateAverage(time_delays);
+    double delay_variance = CalculateVariance(time_delays, delay_average);
+
+    std::chrono::nanoseconds ns_in_one_second = std::chrono::seconds(1);
+    double expected_delay_average =
+        ns_in_one_second.count() / static_cast<double>(request_rate);
 
-    void CheckSequences(PerfAnalyzerParameters params) {
-      // TODO: Add actual checks
-      auto stats = GetStats();
+    if (request_distribution == POISSON) {
+      // By definition, variance == average for Poisson.
+      //
+      // With such a small sample size for a poisson distribution, there will be
+      // noise. Allow 5% slop
+      //
+      CHECK(
+          delay_average ==
+          doctest::Approx(expected_delay_average).epsilon(0.05));
+      CHECK(delay_variance == doctest::Approx(delay_average).epsilon(0.05));
+    } else if (request_distribution == CONSTANT) {
+      // constant should in theory have 0 variance, but with thread timing
+      // there is obviously some noise.
+      //
+      // Allow it to be at most 5% of average
+      //
+      auto max_allowed_delay_variance = 0.05 * delay_average;
+
+      // Constant should be pretty tight. Allowing 1% slop there is noise in the
+      // thread scheduling
+      //
+      CHECK(
+          delay_average ==
+          doctest::Approx(expected_delay_average).epsilon(0.01));
+      CHECK_LT(delay_variance, max_allowed_delay_variance);
+    } else {
+      throw std::invalid_argument("Unexpected distribution type");
     }
+  }
 
-    std::vector<int64_t> GatherTimeBetweenRequests(const std::vector<std::chrono::time_point<std::chrono::system_clock>>& timestamps) {
-      std::vector<int64_t> time_between_requests;
+  void CheckSequences(PerfAnalyzerParameters params)
+  {
+    auto stats = GetStats();
```

Comment:
Sometime before here and elsewhere where `GetStats()` is called, should the load manager be stopped? It seems like the stats object (`std::shared_ptr<cb::MockClientStats>`) is going to continue to be updated as its values are verified.

Reply:
If I stop it, then all of the in-flight sequences are forced to finish, which ends up with sequences that are shorter than the expected length.  For now it is probably best to keep it running. If a request happens while we are checking it shouldn't cause any failures.

Reply:
Can you leave a `// FIXME: <description>` in the code?

Reply:
I killed the manager for some non-sequence tests, and added a fixme into TestSequences